### PR TITLE
Add file caching [databricks]

### PR DESCRIPTION
**Pull request by jlowe**
This adds a file caching solution to help accelerate repeated GPU reads to data on remote filesystems.  It is disabled by default, but can be enabled by setting spark.rapids.filecache.enabled=true.  It leverages the same local directories that are configured for the Spark application, and by default will use up to half of the available disk space of those directories for the filecache.  The size can be explicitly specified by setting spark.rapids.filecache.maxBytes.  Only GPU-accelerated Parquet reads are supported by the filecache.

### CODE REVIEW

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileSourceScanExec.scala [hunk 1]
```diff
@@ -428,7 +429,7 @@ case class GpuFileSourceScanExec(
     BUFFER_TIME -> createNanoTimingMetric(MODERATE_LEVEL, DESCRIPTION_BUFFER_TIME),
     FILTER_TIME -> createNanoTimingMetric(DEBUG_LEVEL, DESCRIPTION_FILTER_TIME),
     PEAK_DEVICE_MEMORY -> createSizeMetric(MODERATE_LEVEL, DESCRIPTION_PEAK_DEVICE_MEMORY)
-  ) ++ {
+  ) ++ createFileCacheMetrics() ++ {
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileSourceScanExec.scala:432]: Do we want to add the metrics only if this is for parquet??  At least for now?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 2]
```diff
@@ -536,12 +540,36 @@ private case class GpuParquetFileFilterHandler(@transient sqlConf: SQLConf) {
...
+        hmb.slice(MAGIC.length, hmb.getLength - Integer.BYTES - MAGIC.length)
+      }
+    }.getOrElse {
+      metrics.getOrElse(GpuMetric.FILECACHE_FOOTER_MISSES, NoopMetric) += 1
+      withResource(readFooterBuffer(filePath, conf)) { hmb =>
+        metrics.getOrElse(GpuMetric.FILECACHE_FOOTER_MISSES_SIZE, NoopMetric) += hmb.getLength
+        // footer was not cached, so try to cache it
+        val cacheToken = FileCache.get.startFooterCache(filePathString)
+        cacheToken.foreach { t =>
+          t.complete(hmb.slice(0, hmb.getLength))
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:562]: I don't understand what the point of this is. You get back a cacheToken, and then need to call complete on it?  Some comments here would really be helpful. Later code makes it clearer what is happening, but still some comments would be helpful.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 3]
```diff
@@ -1396,39 +1446,136 @@ trait ParquetPartitionReaderBase extends Logging with ScanWithMetrics
...
+    val startPos = out.getPos
+    val filePathString: String = filePath.toString
+    val remoteItems = new ArrayBuffer[CopyRange](blocks.length)
+    var totalBytesToCopy = 0L
+    withResource(new ArrayBuffer[LocalCopy](blocks.length)) { localItems =>
+      blocks.foreach { block =>
+        block.getColumns.asScala.foreach { column =>
+          val columnSize = column.getTotalSize
+          val outputOffset = totalBytesToCopy + startPos
+          val channel = FileCache.get.getDataRangeChannel(filePathString,
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1463]: Do we need to worry about having too many channels open at once?  Possibly in a follow on issue?

**Review by revans2 — COMMENTED** [review]
Generally looks good, but I did have a few questions and nits.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala (hunk 3 above)
Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1463]: Chatted with @revans2 offline and this is not an issue.

**Comment by jlowe** [comment]
build

**Review by revans2 — APPROVED** [review]

**Review by tgravescs — DISMISSED** [review]

**Comment by jlowe** [comment]
CI had a hiccup downloading dependencies, rekicking.

**Comment by jlowe** [comment]
build

**Comment by jlowe** [comment]
build

**Review by tgravescs — APPROVED** [review]

**Pull request merged**